content/what-is-cross-origin-resource-sharing/index.md

```diff
@@ -0,0 +1,166 @@
+
+While wildcards will work for simple requests (requests without [HTTP cookies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies) or HTTP authentication information), requests with credentials will often encounter a CORS not authorized error.
+
+That's because in requests with credentials (cookies), the wildcard is treated as the literal method name or origin name "*" without special semantics. This occurs both in `Access-Control-Allow-Origin` and `Access-Control-Allow-Methods`. And some browsers like Safari simply don't have support for wildcards at all, simple request or not.
+
+All in all, it's good hygiene to avoid the wildcard and use a comma-separated list when configuring CORS.
+
+## CORS vulnerability
+
+When configured improperly, CORS can also lead to major vulnerabilities. Below, we'll list a couple of common issues when configuring CORS.
+
+**Mishandling origin whitelist**
+
+One of the easiest mistakes to make when implementing CORS is mishandling the origin whitelist. When whitelisting origins, it's often easy to do simple matches with URL prefixes or suffixes, or using regular expressions. However this can lead to quite a few issues. Let's look at some examples below.
+
+Let's say that we grant access to all websites with the prefix `whitelisted-website.com`. This helps us grant access to origins such as `whitelisted-website.com/blog` or `whitelisted-website.com/login`.
+
+But a malicious actor might come around and use the origin - `whitelisted-website.com.malicious.com` and bypass our security measures.
+
+Another approach that we could use is to grant all websites with the suffix `whitelisted-website.com`. This makes it easy for us to grant access to `api.whitelisted-website.com`.
+
+But an attacker could use a website such as `maliciouswhitelisted-website.com` and gain access.
+
+The best approach here to avoid potential abuse is to explicitly define origins on the whitelist for sensitive operations when implementing CORS - for example, specify the string `https://whitelisted-website.com` which will grant access to all paths on that domain.
+
+**Requests with null origin**
+
+Another common misconfiguration is whitelisting origins with the value null. Browsers might send the value null in the origin header in situations such as:
+
+-   Request with file:
+-   Sandboxed cross-origin requests
+
+In this case, an attacker can use various tricks to generate a request containing the value null as the origin which is whitelisted in our configurations. For example, the attacker could use the following sandboxed iframe exploit -
+
+```html
+<iframe src="data:text/html" sandbox="allow-scripts allow-top-navigation allow-forms allow-same-origin">
+function reqlistener() { console.log(this.responseText) }
+var req = new XMLHttpRequest();
+req. = reqlistener();
+req.open("GET", 'vulnerable.com/sensitive', true);
+req.withCredentials = true;
+req.send();
```

Comment:
```suggestion
<iframe src="data:text/html" sandbox="allow-scripts allow-top-navigation allow-forms allow-same-origin">
   function reqlistener() {
      console.log(this.responseText)
   }
   var req = new XMLHttpRequest();
   req.onload = reqlistener;
   req.open("GET", 'vulnerable.com/sensitive', true);
   req.withCredentials = true;
   req.send();
```

Reply:
The code snippet you had added was wrong and not properly indented. Please make sure mistakes like this don't happen again